auth/certificate_authenticator.cc

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ *
+ */
+
+/*
+ * SPDX-License-Identifier: (AGPL-3.0-or-later and Apache-2.0)
+ */
+
+#include "auth/certificate_authenticator.hh"
+
+#include <regex>
+
+#include "utils/class_registrator.hh"
+#include "replica/database.hh"
+#include "cql3/query_processor.hh"
+#include "db/config.hh"
+
+static const auto CERT_AUTH_NAME = "com.scylladb.auth.CertificateAuthenticator";
+const std::string_view auth::certificate_authenticator_name(CERT_AUTH_NAME);
+const std::string subject_key = "subject";
+
+static logging::logger clogger("certificate_authenticator");
+
+static const class_registrator<auth::authenticator
+    , auth::certificate_authenticator
+    , cql3::query_processor&
+    , ::service::migration_manager&> cert_auth_reg(CERT_AUTH_NAME);
+
+auth::certificate_authenticator::certificate_authenticator(cql3::query_processor& qp, ::service::migration_manager&)
+    : _subject_expr([&] {
+        auto& conf = qp.db().get_config();
+        auto query = conf.auth_certificate_role_query();
+
+        if (query.empty()) {
+            throw std::invalid_argument("No role extraction query specified.");
```

Comment:
Can we provide a reasonable default?

Reply:
If someone suggests one. My best choice would be `CN=(\w+)`, i.e. CNAME = role. But I will let those that asked for the feature chime in before deciding anything.

Reply:
@tzach @tnozicka

Reply:
@tzach @tnozicka please comment on default regex to extract role from certificate.

Reply:
btw, what about security? Can a user specify some other property (let's say city) as CN=superuser and pretend to be the superuser?

Reply:
He can, but since we only accept the certificate subject iff the certificate signature itself verifies against the configured trust store, he would also have to steal/have the signing certificates used to issue the accepted certs. Which I guess is equivalent to knowing a normal password. But harder (?).

The only strengthening I can think of might be useful is perhaps restricting the allowed signers of super user certificates further, though that becomes tricky, both w.r. configuring and also querying the info from gnutls.

Reply:
The attacker may send a certificate to be signed to the security administrator, and get it signed if the security administrator doesn't look at the other fields.

Reply:
I assume certificates and id:s are not distributed like that. It would be the equivalent of handing out passwords on demand. The idea is for the admin (scylla serverless team) to assign an ID to a user/user group, sign a cert with the proper subject, and he then uses it.

But again, if we worry about specificly pretending to be super user, having a more complex trust setup (either subject mask for signing cert or separate trust store) for super user would at least prevent that. 
However, again, for a multi-tenant setup, pretending to be a different user (non-super user) would be just as bad (almost), so I think it is a problem that can't really be 100% solved. Signing authority need to be careful when issuing certs, Same as any other auth.

Reply:
I don't know enough about how certificates are issued to say. What's the existing practice?

/cc @Lorak-mmk in case he knows.

Reply:
I would assume it depends on issuer/org. But for this use case, I still assume that a central authority (sysadm) creates roles + cert + signs it with his root cert, then send the result to the client which will connect.

Reply:
@tzach @tnozicka still need input here

Reply:
Why the "certificate_property" stuff? A simple regex handles this (selecting different attributes). I.e. `auth_certificate_role_query: (?:CN=([^,\s]+))|(?:SAN=([^,\s]+))`.  (You even provide an example below)

Reply:
SAN as source might be problematic. Not to mention a misuse of information, since it should be limited to DNS names...

Reply:
... Well, email and other... I guess it could be valid. Starting to get your request on which field to query. Huge downside in that this requires changes to seastar first though, since we have no way to get alt names right now.

Reply:
regarding the cert safety, the client CA is preferably a dedicated one (that's why most people don't need the regex parsing as they control the CN used for signing) but even with the mapping both client CA and the mapping is controlled by an administrator a so a user can't ask to validate against a different fields to forfeit an identity. Usually it won't come to SANs either as there is not really a need for domains with dedicated client CA.

Reply:
Well, you still need regex (or similar) to extract the relevant role name from a subject (which is a map of values, CN=... etc). While I am not sold on allowing SAN as well, the argument was that normal subject length is limited, and @Ten0  wants potentially longer role names (IIUC?). I can see the value in this, though it seems like a bit of a stretched issue... Just use shorter role names. :-) But I think his notion is using "real" certificates, not special, home-brewed auth-only ones, which might be a use case.

Reply:
>> Can we provide a reasonable default?
> My best choice would be CN=(\w+), i.e. CNAME = role

I agree this should but an advanced option and by default it should just use the CN

Reply:
That is literally "just use the CN", since CN is an attribute in the subject string. Note that we are _not_ breaking the subject up into attribute pairs, we treat it as a fully matched string. CN=(\w) mean literally "take the (word-only) string of CN attribute).

Reply:
It's not that want longer role names, it's that I want to authenticate based on host names (e.g give role `my-service` to somebody who has a cert for `my-service--instance-12.prod.my-example-company.com`), and host names don't fit in CN.

Reply:
Ah, ok, that kinda makes sense. :-)